test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,536 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "testBase.h"
+#include "harness/testHarness.h"
+#include <vector>
+
+/* Negative Tests for clGetDeviceInfo */
+int test_negative_get_device_info(cl_device_id deviceID, cl_context context,
+                                  cl_command_queue queue, int num_elements)
+{
+
+    cl_device_type device_type = 0;
+    cl_int err = clGetDeviceInfo(nullptr, CL_DEVICE_TYPE, sizeof(device_type),
+                                 &device_type, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_DEVICE,
+        "clGetDeviceInfo should return CL_INVALID_DEVICE when: \"device is not "
+        "a valid device\" with a nullptr",
+        TEST_FAIL);
+
+    err =
+        clGetDeviceInfo(reinterpret_cast<cl_device_id>(context), CL_DEVICE_TYPE,
+                        sizeof(device_type), &device_type, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_DEVICE,
+        "clGetDeviceInfo should return CL_INVALID_DEVICE when: \"device is not "
+        "a valid device\" with a with a valid object which is NOT a device",
+        TEST_FAIL);
+
+    constexpr cl_device_info INVALID_PARAM_VALUE = 0;
+    err = clGetDeviceInfo(deviceID, INVALID_PARAM_VALUE, 0, nullptr, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clGetDeviceInfo should return CL_INVALID_VALUE when: \"param_name is "
+        "not one of the supported values\"",
+        TEST_FAIL);
+
+    err = clGetDeviceInfo(deviceID, CL_DEVICE_TYPE, 0, &device_type, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clGetDeviceInfo should return CL_INVALID_VALUE when: \"size in bytes "
+        "specified by param_value_size is < size of return type and "
+        "param_value is not a NULL value\"",
+        TEST_FAIL);
+
+    return TEST_PASS;
+}
+
+/* Negative Tests for clGetDeviceIDs */
+int test_negative_get_device_ids(cl_device_id deviceID, cl_context context,
+                                 cl_command_queue queue, int num_elements)
+{
+    cl_platform_id platform = getPlatformFromDevice(deviceID);
+
+    cl_device_id devices = nullptr;
+    cl_int err =
+        clGetDeviceIDs(nullptr, CL_DEVICE_TYPE_DEFAULT, 1, &devices, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_PLATFORM,
+        "clGetDeviceIDs should return CL_INVALID_PLATFORM when: \"platform is "
+        "not a valid platform\" using a nullptr",
```

Comment:
Passing `nullptr` for the platform is allowed and the behaviour is implementation-defined:

>platform refers to the platform ID returned by clGetPlatformIDs or can be NULL. If platform is
NULL, the behavior is implementation-defined.

Reply:
> Passing `nullptr` for the platform is allowed and the behaviour is implementation-defined:
> 
> > platform refers to the platform ID returned by clGetPlatformIDs or can be NULL. If platform is
> > NULL, the behavior is implementation-defined.

Then we have merged incorrect test case [here ](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/api/negative_platform.cpp#L45)? Related with question #1222

Reply:
Yes, that code is incorrect. I've replied on the issue and asked @chemis01 to prepare a fix.